microceph/ceph/nfs.go

```diff
@@ -0,0 +1,316 @@
+		service := "nfs"
+		filter := database.GroupedServiceFilter{
+			Service: &service,
+			GroupID: &clusterID,
+		}
+		groupedServices, err := database.GetGroupedServices(ctx, tx, filter)
+		if err != nil {
+			return fmt.Errorf("failed to get grouped services records: %w", err)
+		}
+
+		if len(groupedServices) > 0 {
+			// There's still at least one GroupedService referencing this ServiceGroup.
+			return nil
+		}
+
+		// Delete the ServiceGroup record.
+		err = database.DeleteServiceGroup(ctx, tx, "nfs", clusterID)
+		if err != nil {
+			return fmt.Errorf("failed to delete service group record: %w", err)
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+// createNFSServiceGroupRecord creates a record in the grouped_services database.
+func createNFSServiceGroupRecord(ctx context.Context, s interfaces.StateInterface, clusterID, info string) error {
+	if s.ClusterState().ServerCert() == nil {
+		return fmt.Errorf("no server certificate")
+	}
+
+	err := s.ClusterState().Database().Transaction(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		// Create the GroupedService record.
+		_, err := database.CreateGroupedService(ctx, tx, database.GroupedService{Member: s.ClusterState().Name(), GroupID: clusterID, Service: "nfs", Info: info})
+		if err != nil {
+			return fmt.Errorf("failed to record grouped service: %w", err)
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+// deleteNFSServiceGroupRecord removes a record in the grouped_services database.
+func deleteNFSServiceGroupRecord(ctx context.Context, s interfaces.StateInterface, clusterID string) error {
+        if s.ClusterState().ServerCert() == nil {
+                return fmt.Errorf("no server certificate")
+        }
+
+        err := s.ClusterState().Database().Transaction(ctx, func(ctx context.Context, tx *sql.Tx) error {
+                // Delete the GroupedService record.
+                err := database.DeleteGroupedService(ctx, tx, s.ClusterState().Name(), "nfs", clusterID)
+                if err != nil {
+                        return fmt.Errorf("failed to delete grouped service record: %w", err)
+                }
+
+                return nil
+        })
+
+        return err
+}
+
+// startNFS starts the NFS service.
+func startNFS() error {
+	err := snapStart("nfs-ganesha", true)
+	if err != nil {
+		return fmt.Errorf("failed to start NFS Ganesha service: %w", err)
+	}
+
+	return nil
+}
+
+// stopNFS stops the NFS service.
+func stopNFS() error {
+	err := snapStop("nfs-ganesha", true)
+	if err != nil {
+		return fmt.Errorf("failed to stop NFS Ganesha service: %w", err)
+	}
+
+	return nil
+}
+
+// createNFSKeyring creates the NFS keyring.
+func createNFSKeyring(path, clusterID string) error {
+	if err := os.MkdirAll(path, 0770); err != nil {
+		return err
+	}
+	// Create the keyring.
+	keyringPath := filepath.Join(path, "keyring")
+	if _, err := os.Stat(keyringPath); err == nil {
+		return nil
+	}
+
+	err := genAuth(
+		keyringPath,
+		"client.nfs.ganesha",
```

Comment:
The client keyring is not parametrized with `hostname` or `cluster-id`. This means that when we will get the same client key for all nfs-ganesha daemons in all nfs clusterids. Should be of the form `client.nfs-<clusterid>.<hostname>` ensuring each deamon gets a unique ceph client key.